src/dataflow/source/kinesis.rs

```diff
@@ -70,7 +70,8 @@ where
         let activator = scope.activator_for(&info.address[..]);
 
         move |cap, output| {
-            let (client, shard_iterator) = match &mut state {
+            // todo@jldlaughlin: We need to be able to update our list of shards dynamically. #2607
```

Comment:
A way to achieve this fairly easily might be to call kinesis_client.list_shards (=list_shard) at the beginning (outside of the closure) and store the list of "known" shards. If the timestamp_history.get(source_id).len() > list_shards.len() then you know that a new partition has been added and it's worth calling kinesis_client.list_shards again? 

This should work even if shards get deleted as long as we use kinesis_client.list_shards to update the list of shards with new ids and don't remove ones that have been deleted.